controllers/progressiverollout_controller.go

```diff
@@ -61,5 +69,93 @@ func (r *ProgressiveRolloutReconciler) Reconcile(ctx context.Context, req ctrl.R
 func (r *ProgressiveRolloutReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	return ctrl.NewControllerManagedBy(mgr).
 		For(&deploymentskyscannernetv1alpha1.ProgressiveRollout{}).
+		Watches(
+			&source.Kind{Type: &argov1alpha1.Application{}},
+			handler.EnqueueRequestsFromMapFunc(r.requestsForApplicationChange)).
+		Watches(
+			&source.Kind{Type: &corev1.Secret{}}, handler.EnqueueRequestsFromMapFunc(r.requestsForSecretChange)).
 		Complete(r)
 }
+
+func (r *ProgressiveRolloutReconciler) requestsForApplicationChange(o client.Object) []reconcile.Request {
+
+	/*
+		We trigger a reconciliation loop on an Application event if:
+		- the Application owner is referenced by a ProgressiveRollout object
+	*/
+
+	var requests []reconcile.Request
+	var list deploymentskyscannernetv1alpha1.ProgressiveRolloutList
+	ctx := context.Background()
+
+	app, ok := o.(*argov1alpha1.Application)
+	if !ok {
+		err := fmt.Errorf("expected application, got %T", o)
+		r.Log.Error(err, "failed to convert object to application")
+		return nil
+	}
+
+	if err := r.List(ctx, &list); err != nil {
+		r.Log.Error(err, "failed to list ProgressiveRollout")
+		return nil
+	}
+
+	for _, pr := range list.Items {
+		if pr.HasOwnerReference(app.GetOwnerReferences()) {
+			requests = append(requests, reconcile.Request{NamespacedName: types.NamespacedName{
+				Namespace: pr.Namespace,
+				Name:      pr.Name,
+			}})
+		}
+	}
+
+	return requests
+}
+
+func (r *ProgressiveRolloutReconciler) requestsForSecretChange(o client.Object) []reconcile.Request {
+
+	/*
```

Comment:
Just so I'm clear; is the logic of reconciling after a secret existing because we want to notify the current deployments that a cluster has been marked as drained?
Just trying to understand the impact, would that effectively trigger a deployment? Idempotency should mean no change, but just concious that it could mean _every single_ progressiverollout will trigger at the same time?

There might be something I'm not getting here.

Reply:
We trigger a reconciliation loop every time there is a change to a secret object in Lubernetes. This is why we filter using this function, because we don't care about *every* secrets.

Let's say we add a new label to a secret, this will trigger a reconciliation loop for every ProgressiveRollout targeting that cluster.

We could minimize the impact and do something like "don't forward the event if we are *adding* the following `key:value`". In that case we could avoid triggering the reconciliation loop when we add `drained: true` but still trigger it when we remove it or change to `drained: false`.

If that is what you mean, I think I will raise a separate issue to fix that. We might want by default filter out all the labels used in the (future) requeue work.

Reply:
that is what I meant indeed. Happy to create a separate issue.

Reply:
> The question still stands when a new cell is added, in this case the argocd secret would be updated too, does this mean all the progressive rollouts would trigger a deployment?

Only if there are Applications targeting that new cluster, so none until we have Applications able to deploy there.

We can use with https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.8.2/pkg/predicate to create our own filter logic so we send only the events we care about.